src/execution_plan/execution_plan.c

```diff
@@ -672,20 +672,20 @@ static ExecutionPlanSegment *_NewExecutionPlanSegment(RedisModuleCtx *ctx, Graph
 	if(prev_op) {
 		// Need to connect this segment to the previous one.
 		// If the last operation of this segment is a potential data producer, join them
-		// under an Apply operation. (This could be a Cartesian Product if preferred.)
+		// under a Cartesian Product operation. (This could be an Apply op if preferred.)
 		if(parent_op->type & OP_TAPS) {
-			OpBase *op_apply = NewApplyOp();
+			OpBase *op_cp = NewCartesianProductOp();
 			uint prev_projection_count = array_len(prev_projections);
 			// The Apply op must be associated with the IDs projected from the previous segment
 			// in case we're placing filter operations.
 			// (although this is a rather ugly way to accomplish that.)
-			op_apply->modifies = array_new(uint, prev_projection_count);
+			op_cp->modifies = array_new(uint, prev_projection_count);
 			for(uint i = 0; i < prev_projection_count; i ++) {
-				op_apply->modifies = array_append(op_apply->modifies, i);
+				op_cp->modifies = array_append(op_cp->modifies, i);
```

Comment:
Can you shad some light on the modifies array?

Reply:
Sure! It's unchanged from master except that it now stores Record IDs rather than aliases (and is an array instead of a Vector).

When choosing the optimal position to put Filter ops, we traverse the operation chain until every Record ID specified in a filter tree (`FilterTree_CollectModified`) has been seen in a `modifies` array (`ExecutionPlan_LocateReferences`).

Record IDs are scoped to an ExecutionPlanSegment, and WITH projections are the first _n_ Record IDs in the subsequent segment. (Which is the reason for this logic.)

Since Cypher allows aliases to be reused in different WITH-separated segments, we would need a fix like this even if we were using aliases instead of IDs.